dill/session.py

```diff
@@ -0,0 +1,220 @@
+            continue
+        if '.' not in modname:
+            modmap.top_level[id(module)] = modname
+        for objname, modobj in module.__dict__.items():
+            modmap.by_name[objname].append((modobj, modname))
+            modmap.by_id[id(modobj)].append((modobj, objname, modname))
+    return modmap
+
+def _lookup_module(modmap, name, obj, main_module):
+    """lookup name or id of obj if module is imported"""
+    for modobj, modname in modmap.by_name[name]:
+        if modobj is obj and sys.modules[modname] is not main_module:
+            return modname, name
+    if isinstance(obj, SESSION_IMPORTED_AS_TYPES):
+        for modobj, objname, modname in modmap.by_id[id(obj)]:
+            if sys.modules[modname] is not main_module:
+                return modname, objname
+    return None, None
+
+def _stash_modules(main_module):
+    modmap = _module_map()
+    newmod = ModuleType(main_module.__name__)
+
+    imported = []
+    imported_as = []
+    imported_top_level = []  # keep separeted for backwards compatibility
+    original = {}
+    for name, obj in vars(main_module).items():
+        if obj is main_module:
+            original[name] = newmod  # self-reference
+            continue
+
+        # Avoid incorrectly matching a singleton value in another package (ex.: __doc__).
+        if any(obj is singleton for singleton in (None, False, True)) or \
+                isinstance(obj, ModuleType) and _is_builtin_module(obj):  # always saved by ref
+            original[name] = obj
+            continue
+
+        source_module, objname = _lookup_module(modmap, name, obj, main_module)
+        if source_module:
+            if objname == name:
+                imported.append((source_module, name))
+            else:
+                imported_as.append((source_module, objname, name))
+        else:
+            try:
+                imported_top_level.append((modmap.top_level[id(obj)], name))
+            except KeyError:
+                original[name] = obj
+
+    if len(original) < len(main_module.__dict__):
+        newmod.__dict__.update(original)
+        newmod.__dill_imported = imported
+        newmod.__dill_imported_as = imported_as
+        newmod.__dill_imported_top_level = imported_top_level
+        return newmod
+    else:
+        return main_module
+
+def _restore_modules(unpickler, main_module):
+    try:
+        for modname, name in main_module.__dict__.pop('__dill_imported'):
+            main_module.__dict__[name] = unpickler.find_class(modname, name)
+        for modname, objname, name in main_module.__dict__.pop('__dill_imported_as'):
+            main_module.__dict__[name] = unpickler.find_class(modname, objname)
+        for modname, name in main_module.__dict__.pop('__dill_imported_top_level'):
+            main_module.__dict__[name] = __import__(modname)
+    except KeyError:
+        pass
+
+def _filter_objects(main, exclude_extra, include_extra, obj=None):
```

Comment:
I don't like the names `exclude_extra` and `include_extra`.  Should their defaults be `None`?  Is the name `main` better as `namespace`?  Also, it's not obvious if `obj` is even a good name, as it's just passed to `filter_namespace` (and there's currently no docs here).

Reply:
> I don't like the names `exclude_extra` and `include_extra` (better is just `exclude` and `include`).
I think they were like you suggested initially. I appended the 'extra' to emphasize that they are added to the filters from settings. OK, I'll change them back to match the `dump_session()` parameter and annotate this in the function docstring.

> Should their defaults be `None`?
Isn't it better to leave them as required arguments instead of optional, as it is an internal function with a single use case?

> Is the name `main` better as `namespace`?
I used more generic names with the filtering framework in `_utils.py` because it may be used for other things not related to module variables. In `session.py` I tried to be consistent with the naming in other functions. Both seem fine for me.

> Also, it's not obvious if `obj` is even a good name, as it's just passed to `filter_namespace` (and there's currently no docs here).
If you are more or less OK with the implementation, I'll start writing the documentation. I borrowed this `func(..., obj=obj)` notation from `Pickler.save_reduce()`. This parameter is used to avoid creating a second module object after `_stash_modules()` already created one. But maybe this is a case of premature optimization, and I think the filtering should actually run first...

> And... I hate to harp on the naming, but is `_filter_objects` really an appropriate name for this function?
Naming is important. `_filter_variables`? 🐧

Reply:
> I'll change them back to match the dump_session() parameter and annotate this in the function docstring.

Great.

> Isn't it better to leave them as required arguments instead of optional, as it is an internal function with a single use case?

By this logic, `obj` should not have a default.

>  I used more generic names with the filtering framework in _utils.py because it may be used for other things not related to module variables. In session.py I tried to be consistent with the naming in other functions. Both seem fine for me.

Right... I've been thinking that with some of the recent changes, this code probably is useful beyond a "session", and definitely with your recent PRs, the module doesn't need to be "main" (and is generalized to "any" module).  I'm thinking that we should adjust the names accordingly.

> If you are more or less OK with the implementation, I'll start writing the documentation. I borrowed this func(..., obj=obj) notation from Pickler.save_reduce(). 

I am more or less OK with the implementation.  Yeah, it's obvious wha the `obj` is for, and since it's an internal function, probably fine to leave it as is.  I'm guessing it could be avoided, and the interface improved... but we can save that discussion for another time.

> _filter_variables?

No. It's applying filters to whatever "main" is.  How about `_apply_filters`?

Reply:
About naming... The parameter `obj` is gone as this is now called before `_stash_modules()` (and it was an unnecessary micro-optimization). I ended using `_apply_filters()` to name a new private function of the `FilterRules` class, so that I was able to unroll the exclude/include filtering loop. The current calling sequence and function/method signatures (not including default values of `dump_module`) are:
```python
dump_module(filename, module, *, refimported, exclude, include, base_rules)
--> _filter_vars(main_module, exclude, include, base_rules)
    --> FilterRules.filter_vars(namespace)
        --> FilterRules._apply_filters(filter_set, objects)
```

Is it better to just call everything `(_)apply_filters`?

Reply:
It's often nice... and generally done when (1) it has the same signature or (2) one is an obvious support function for the other (i.e. `foo` and `_foo`).  It appears that (1) doesn't apply, but (2) may.  I think that using `FilterRules.apply_filters(namespace)` would be consistent.